rotkehlchen/chain/evm/decoding/paraswap/decoder.py

```diff
@@ -0,0 +1,196 @@
+
+from .constants import BUY_SIGNATURE, CPT_PARASWAP, SWAP_SIGNATURE as PARASWAP_SWAP_SIGNATURE
+
+if TYPE_CHECKING:
+    from rotkehlchen.chain.evm.decoding.base import BaseDecoderTools
+    from rotkehlchen.chain.evm.node_inquirer import EvmNodeInquirer
+    from rotkehlchen.user_messages import MessagesAggregator
+
+
+class ParaswapCommonDecoder(DecoderInterface):
+    def __init__(
+            self,
+            evm_inquirer: 'EvmNodeInquirer',
+            base_tools: 'BaseDecoderTools',
+            msg_aggregator: 'MessagesAggregator',
+            router_address: ChecksumEvmAddress,
+            fee_receiver_address: ChecksumEvmAddress,
+    ) -> None:
+        super().__init__(evm_inquirer, base_tools, msg_aggregator)
+        self.router_address = router_address
+        self.fee_receiver_address = fee_receiver_address
+        logger = logging.getLogger(__name__)
+        self.log = RotkehlchenLogsAdapter(logger)
+
+    def _decode_swap(
+            self,
+            context: DecoderContext,
+            receiver: ChecksumEvmAddress,
+            sender: ChecksumEvmAddress,
+    ) -> DecodingOutput:
+        """This function is used to decode the swap done by paraswap.
+        At the moment it does not decode the fee in ETH. For example:
+        etherscan.io/tx/0xeeea20b39f157fe59fa4904fd4b62f8971188b53d05c6831e0ed67ee157e40c2"""
+        if not self.base.is_tracked(sender) and not self.base.is_tracked(receiver):
+            return DEFAULT_DECODING_OUTPUT
+
+        out_event = in_event = returned_event = None
+        for event in context.decoded_events:
+            if (
+                event.event_type == HistoryEventType.TRADE and
+                event.event_subtype == HistoryEventSubType.SPEND or
+                event.event_type == HistoryEventType.SPEND and
+                event.event_subtype == HistoryEventSubType.NONE and
+                event.location_label == sender
+            ):  # find the send event
+                event.counterparty = CPT_PARASWAP
+                event.event_type = HistoryEventType.TRADE
+                event.event_subtype = HistoryEventSubType.SPEND
+                event.address = self.router_address
+                out_event = event
+            elif event.event_type == HistoryEventType.RECEIVE and event.location_label == sender:  # find the receive event  # noqa: E501
+                event.event_type = HistoryEventType.TRADE
+                event.event_subtype = HistoryEventSubType.RECEIVE
+                event.counterparty = CPT_PARASWAP
+                event.notes = f'Receive {event.balance.amount} {event.asset.resolve_to_asset_with_symbol().symbol} as the result of a swap in paraswap'  # noqa: E501
+                event.address = self.router_address
+                # use this index as the event may be a native currency transfer
+                # and appear at the start
+                event.sequence_index = context.tx_log.log_index
```

Comment:
Don't you need to check if the spend and receive have the sequence index in the correct sequence? it could be that a previous logic replaces the sequence and you end up with an undesired order. We have `get_sequence_index` that I don't know if might be useful for you in this case. Just letting you know

Reply:
I see, I'm removing this line and then using `maybe_reshuffle_events` to order them.